File: Problems/Algorithms/36. Valid Sudoku/ValidSudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        for (int i = 0; i < 9; i++) {
            if (!isValidRow(board, i)) {
                return false;
            }
            if (!isValidCol(board, i)) {
                return false;
            }
        }

        for (int i = 0; i < 9; i += 3) {
            for (int j = 0; j < 9; j += 3) {
                if (!isValidBox(board, i, j)) {
                    return false;
                }
            }
        }

        return true;
    }

    private boolean isValidRow(char[][] board, int row) {
        boolean[] used = new boolean[9];
        for (int i = 0; i < 9; i++) {
            if (board[row][i] != '.') {
                if (used[board[row][i] - '1']) {
                    return false;
                }
                used[board[row][i] - '1'] = true;
            }
        }
        return true;
    }

    private boolean isValidCol(char[][] board, int col) {
        boolean[] used = new boolean[9];
        for (int i = 0; i < 9; i++) {
            if (board[i][col] != '.') {
                if (used[board[i][col] - '1']) {
                    return false;
                }
                used[board[i][col] - '1'] = true;
            }
        }
        return true;
    }

    private boolean isValidBox(char[][] board, int row, int col) {
        boolean[] used = new boolean[9];
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (board[row + i][col + j] != '.') {
                    if (used[board[row + i][col + j] - '1']) {
                        return false;
                    }
                    used[board[row + i][col + j] - '1'] = true;
                }
            }
        }
        return true;
    }
}

File: Problems/Algorithms/36. Valid Sudoku/ValidSudoku.kt
class Solution {
    fun isValidSudoku(board: Array<CharArray>): Boolean {
        for (i in 0..8) {
            if (!isValidRow(board, i)) return false
            if (!isValidCol(board, i)) return false
            if (!isValidBox(board, i)) return false
        }
        
        return true
    }
    
    private fun isValidRow(board: Array<CharArray>, row: Int): Boolean {
        val map: MutableMap<Char, Int> = mutableMapOf<Char, Int>()
        for (j in 0..8) {
            val ch = board[row][j]
            if (ch != '.') {
                if (map.containsKey(ch)) return false
                map[ch] = 1
            }
        }
        
        return true
    }
    
    private fun isValidCol(board: Array<CharArray>, col: Int): Boolean {
        val map: MutableMap<Char, Int> = mutableMapOf<Char, Int>()
        for (i in 0..8) {
            val ch = board[i][col]
            if (ch != '.') {
                if (map.containsKey(ch)) return false
                map[ch] = 1
            }
        }
        
        return true
    }
    
    private fun isValidBox(board: Array<CharArray>, box: Int): Boolean {
        val map: MutableMap<Char, Int> = mutableMapOf<Char, Int>()
        for (i in 0..2) {
            for (j in 0..2) {
                val ch = board[box/3*3+i][box%3*3+j]
                if (ch != '.') {
                    if (map.containsKey(ch)) return false
                    map[ch] = 1
                }
            }
        }
        
        return true
    }
}

